common/src/main/java/com/wynntils/features/user/ChatTabsFeature.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright © Wynntils 2022.
+ * This file is released under AGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.features.user;
+
+import com.google.common.collect.Sets;
+import com.google.gson.reflect.TypeToken;
+import com.wynntils.core.chat.ChatModel;
+import com.wynntils.core.chat.RecipientType;
+import com.wynntils.core.chat.tabs.ChatTab;
+import com.wynntils.core.chat.tabs.ChatTabModel;
+import com.wynntils.core.config.Config;
+import com.wynntils.core.config.TypeOverride;
+import com.wynntils.core.features.UserFeature;
+import com.wynntils.core.managers.Model;
+import com.wynntils.gui.widgets.ChatTabButton;
+import com.wynntils.mc.event.ClientsideMessageEvent;
+import com.wynntils.mc.event.ScreenInitEvent;
+import com.wynntils.mc.event.ScreenRenderEvent;
+import com.wynntils.wynn.event.ChatMessageReceivedEvent;
+import com.wynntils.wynn.event.WorldStateEvent;
+import com.wynntils.wynn.model.WorldStateManager;
+import java.lang.reflect.Type;
+import java.util.Arrays;
+import java.util.List;
+import net.minecraft.client.gui.screens.ChatScreen;
+import net.minecraftforge.eventbus.api.EventPriority;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+
+public class ChatTabsFeature extends UserFeature {
+    public static ChatTabsFeature INSTANCE;
+
+    @Config(visible = false)
+    public List<ChatTab> chatTabs = Arrays.asList(
+            new ChatTab("All", true, null, null),
+            new ChatTab("Global", true, Sets.newHashSet(RecipientType.GLOBAL), null),
+            new ChatTab("Local", true, Sets.newHashSet(RecipientType.LOCAL), null),
+            new ChatTab("Guild", true, Sets.newHashSet(RecipientType.GUILD), null),
+            new ChatTab("Party", true, Sets.newHashSet(RecipientType.PARTY), null),
+            new ChatTab("Private", true, Sets.newHashSet(RecipientType.PRIVATE), null),
+            new ChatTab("Shout", true, Sets.newHashSet(RecipientType.SHOUT), null));
+
+    @TypeOverride
+    private final Type chatTabsType = new TypeToken<List<ChatTab>>() {}.getType();
+
+    @Override
+    public List<Class<? extends Model>> getModelDependencies() {
+        return List.of(ChatModel.class, ChatTabModel.class);
+    }
+
+    // We do this here, and not in ChatTabModel to not introduce a feature-model dependency.
+    @SubscribeEvent(priority = EventPriority.LOWEST)
+    public void onChatReceived(ChatMessageReceivedEvent event) {
+        // Firstly, find the FIRST matching tab with high priority
+        for (ChatTab chatTab : chatTabs) {
+            if (chatTab.isLowPriority()) continue;
+
+            if (chatTab.matchMessageFromEvent(event)) {
+                ChatTabModel.addMessageToTab(chatTab, event.getMessage());
+                return;
```

Comment:
does this mean it's not possible for two tabs that aren't low priority to receive the same message? this seems like a step backwards from legacy functionality

Reply:
Doesn't legacy non low priority consume the message? as in the first matching tab gets to show the message, any other not

Reply:
in legacy that logic only applies to tabs specifically marked as low priority - i had two tabs that were set to receive all messages (one for typing into public chat, one for typing to guild chat), and that structure couldn't work here unless i made all my tabs low priority, which doesn't feel very intuitive 

honestly it might be better to abandon the priority system altogether and implement a different hierarchy system along with "negative" filters, but that could be dealt with in a later PR

Reply:
We could just make consumer tabs. If checked, the specific overlay consumes it's matching messages. How do you feel about that?

Reply:
so there would be consumer/"high priority" tabs, regular tabs, and then low priority? that sounds like a good solution

Reply:
Exactly